### PR TITLE
brought in changes from anomaly experiments

### PR DESCRIPTION
**Pull request by nishaq503**
fix: made things faster

### CODE REVIEW

**Comment by thoward27** [comment]
@nishaq503 you broke a unit test. We cannot merge to master till that's addressed

File: pyclam/manifold.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def __init__(self, manifold: 'Manifold', argpoints: Vector, name: str, **kwargs)
         if not argpoints and self.children:
             self.argpoints = [p for child in self.children for p in child.argpoints]
         elif not argpoints:
-            raise ValueError(f"Cluster {name} needs argpoints.")
+            raise ValueError(f'Cluster {name} needs argpoints: {argpoints}')
```
Comment (thoward27) [inline pyclam/manifold.py]: We already established there are no argpoints, logging them here seems weird

File: pyclam/manifold.py [hunk 2]
```diff
@@ -308,18 +308,29 @@ def partition(self, *criterion) -> Iterable['Cluster']:
             self.manifold.data[farthest],
         ])
 
-        p1_idx, p2_idx = list(), list()
+        p1_idx, p2_idx = [self.argradius], [farthest]
         [(p1_idx if p1 < p2 else p2_idx).append(i)
          for batch in iter(self)
          for i, p1, p2 in zip(batch, *cdist(poles, self.manifold.data[batch], self.metric))]
 
+        # TODO: Neither of these should be possible
```
Comment (thoward27) [inline pyclam/manifold.py]: Then why are we testing for them? bugs elsewhere?

File: pyclam/manifold.py [hunk 3]
```diff
@@ -624,7 +649,7 @@ def build(self, *criterion) -> 'Manifold':
     def build_tree(self, *criterion) -> 'Manifold':
         """ Builds the Cluster-tree. """
         while True:
-            logging.info(f'current depth: {len(self.graphs) - 1}')
+            logging.info(f'current depth: {len(self.graphs) - 1}, {len(self.graphs[-1].clusters.keys())} clusters')
```
Comment (thoward27) [inline pyclam/manifold.py]: Shouldn't need to access keys here

File: pyclam/manifold.py (hunk 2 above)
Comment (nishaq503) [reply pyclam/manifold.py]: I don't know why this invariant breaks. I'll look into it after unifying distance calls.

File: pyclam/manifold.py (hunk 1 above)
Comment (nishaq503) [reply pyclam/manifold.py]: addressed

File: pyclam/manifold.py (hunk 3 above)
Comment (nishaq503) [reply pyclam/manifold.py]: addressed

**Pull request merged**